get_security_annotations.py

```diff
@@ -176,50 +211,45 @@ def log_event_attributes(alert, threat_intel_record, asset, threat_detection, ev
     print(json.dumps(row))
 
 
-def process_events(alert, threat_detection, threat_intel_record, asset, events_iterator, previous_max_event_modified_at, max_event_modified_at):
-    while events_iterator.has_next():
-        event = events_iterator.next()
-
-        modified_at_event_datetime = parse_iso_datetime(event["modifiedAt"])
-        if modified_at_event_datetime > previous_max_event_modified_at:
-            log_event_attributes(alert, threat_intel_record, asset, threat_detection, event)
-
-            if modified_at_event_datetime > max_event_modified_at:
-                max_event_modified_at = modified_at_event_datetime
-
-    return max_event_modified_at
-
-
 def main():
-    # get maximal "modifiedAt" from all processed Sightings stored after previous run
-    previous_max_event_modified_at = read_max_event_modified_at()
-    # variable holding "modifiedAt" maximum for all Sightings currently processed
-    max_event_modified_at = previous_max_event_modified_at
+    # read security events cursor from previous script run (if any) to process just unprocessed or modified events
+    previous_events_end_cursor = read_events_end_cursor()
 
-    alerts_iterator = CollectionIterator("/alert-management/customer/" + CUSTOMER_ID + "/alerts")
+    # iterate over all security events sorted by unique modification sequence number ensuring no event is missed
+    events_iterator = CollectionIterator(build_event_with_threat_detections_ids_url(previous_events_end_cursor))
+    while events_iterator.has_next():
+        event = events_iterator.next()
 
-    # iterate through all alerts
-    while alerts_iterator.has_next():
-        alert = alerts_iterator.next()
+        # get affected asset referred by event
+        affected_asset = get_asset(event["affectedAssetId"])
 
-        threat_detections_iterator = CollectionIterator(build_alert_threat_detections_url(alert["id"]))
+        # get IDs references to threat detections from event (usually one)
+        threat_detection_ids = event["threatDetectionIds"]
 
-        # iterate through all threat_detections
-        while threat_detections_iterator.has_next():
-            threat_detection = threat_detections_iterator.next()
-            asset = get_asset(threat_detection["affectedAssetId"])
-            threat_intel_record = get_threat_intel(threat_detection["threatIntelRecordId"])
+        # process all possible threat detections
+        if len(threat_detection_ids) > 0:
+            # process event with threat detections (called "convicting" security event)
+            for threat_detection_id in threat_detection_ids:
+                # get parent objects to event - threat detections and alert
+                # and also threat intel record from threat-catalog bounded context
+                threat_detection_with_alert_ids = get_threat_detection_with_alert_ids(threat_detection_id)
```

Comment:
It would be better to use bulk operations to get threat detections (also I would rather see to get all threat detections at once and not per event)

Reply:
Reworked to use bulk requests.